brave/src/main/java/brave/propagation/B3Propagation.java

```diff
@@ -13,31 +13,86 @@
- * Implements <a href="https://github.com/apache/incubator-zipkin-b3-propagation">B3
- * Propagation</a>
+ * Implements <a href="https://github.com/openzipkin/b3-propagation">B3 Propagation</a>
  */
 public final class B3Propagation<K> implements Propagation<K> {
+  /** Describes the formats used to inject headers. */
+  enum Format {
+    /** The trace context is encoded with a several fields prefixed with "x-b3-". */
+    MULTI,
+    /** The trace context is encoded with {@link B3SingleFormat#writeB3SingleFormat(TraceContext)}. */
+    SINGLE,
+    /** The trace context is encoded with {@link B3SingleFormat#writeB3SingleFormatWithoutParentId(TraceContext)}. */
+    SINGLE_NO_PARENT
+  }
 
-  public static final Propagation.Factory FACTORY = new Propagation.Factory() {
-    @Override public <K1> Propagation<K1> create(KeyFactory<K1> keyFactory) {
-      return new B3Propagation<>(keyFactory);
+  public static final Propagation.Factory FACTORY = newFactoryBuilder().build();
+
+  public static FactoryBuilder newFactoryBuilder() {
+    return new FactoryBuilder();
+  }
+
+  /**
+   * Defaults to {@link Format#MULTI} for client/server spans and {@link Format#SINGLE_NO_PARENT}
+   * for messaging. Non-request spans default to {@link Format#MULTI}.
+   */
+  public static final class FactoryBuilder {
+    Format injectFormat = Format.MULTI;
+    final EnumMap<Span.Kind, Format[]> kindToInjectFormats = new EnumMap<>(Span.Kind.class);
+
+    FactoryBuilder() {
+      kindToInjectFormats.put(Span.Kind.CLIENT, new Format[] {Format.MULTI});
+      kindToInjectFormats.put(Span.Kind.SERVER, new Format[] {Format.MULTI});
+      kindToInjectFormats.put(Span.Kind.PRODUCER, new Format[] {Format.SINGLE_NO_PARENT});
+      kindToInjectFormats.put(Span.Kind.CONSUMER, new Format[] {Format.SINGLE_NO_PARENT});
     }
 
-    @Override public boolean supportsJoin() {
-      return true;
+    /** Overrides the default format of {@link Format#MULTI}. */
+    public FactoryBuilder injectFormat(Format format) {
+      if (format == null) throw new NullPointerException("format == null");
+      injectFormat = format;
+      return this;
     }
 
-    @Override public String toString() {
-      return "B3PropagationFactory";
+    /** Overrides the injection format used for the indicated {@link Request#kind() span kind}. */
+    public FactoryBuilder injectFormat(Span.Kind kind, Format format) {
+      if (kind == null) throw new NullPointerException("kind == null");
+      if (format == null) throw new NullPointerException("format == null");
+      kindToInjectFormats.put(kind, new Format[] {format});
+      return this;
     }
-  };
+
+    /**
+     * Like {@link #injectFormat}, but writes multiple formats.
+     *
+     * For example, you can set {@link Span.Kind#CLIENT} spans to inject both {@link Format#MULTI}
+     * and {@link Format#SINGLE}, for transition use cases.
+     */
+    public FactoryBuilder injectFormats(Span.Kind kind, Format format1, Format format2) {
```

Comment:
Should this check that both `format1` and `format2` are different and not `SINGLE` and `SINGLE_NO_PARENT`?

Reply:
sgtm